## 4.6.2 Combining & Exporting Data

### This script implements with the following steps:

#### 1. Import the data sets into Jupyter

#### 2. Check the dimensions of the imported dataframes

#### 3. Determine a suitable way to combine the dataframe orders_products_combined with the products data set

#### 4. Confirm the results of the merge using the "merge_flag"

#### 5. Export the newly created dataframe as ords_prods_merge in a suitable format

In [3]:
# Import libraries
import pandas as pd
import numpy as np
import os

### 1. Import the data sets into Jupyter

In [4]:
# Tell Python to remember a main folder path
path = r'/Users/omkargopalakrishnan/Instacart Basket Analysis'

In [5]:
# Import dataset orders_products_combined.pkl
df_ords_prods_combined = pd.read_pickle(os.path.join(path, 'Prepared Data', 'orders_products_combined.pkl'))

In [6]:
# Import dataset products_checked.csv
df_prods = pd.read_csv(os.path.join(path, 'Prepared Data', 'products_checked.csv'), index_col = False)

### 2. Check the dimensions of the imported dataframes

In [7]:
# Check the output
df_ords_prods_combined.head()

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0           0   2539329        1    prior             1          2   
1           0   2539329        1    prior             1          2   
2           0   2539329        1    prior             1          2   
3           0   2539329        1    prior             1          2   
4           0   2539329        1    prior             1          2   

   order_hour_of_day  days_since_prior_order  first_order  product_id  \
0                  8                     NaN         True         196   
1                  8                     NaN         True       14084   
2                  8                     NaN         True       12427   
3                  8                     NaN         True       26088   
4                  8                     NaN         True       26405   

   add_to_cart_order  reordered _merge  
0                  1          0   both  
1                  2          0   both  
2                  3          0   both  
3                  4          0   both  
4                  5          0   both

In [8]:
df_ords_prods_combined.shape

(32434489, 13)

In [9]:
# Check the output
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [10]:
# Drop unnecessary columns
df_prods = df_prods.drop(['Unnamed: 0'], axis=1)

In [11]:
# Check the output
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [14]:
df_prods.shape

(49672, 5)

### 3. Determine a suitable way to combine the orders_products_combined dataframe with the products data set¶

#### Despite their different shape, the dataframes df_ords_prods_combined and df_prods dataframe can be combined on their shared column "product_id". For the purpose of this project, a default inner join is advisable.

In [20]:
df_ords_prods_merge = df_prods.merge(df_ords_prods_combined, on='product_id', indicator='merge_flag')

#### When executed using the script df_ords_prods_merge = df_prods.merge(df_ords_prods_combined, on = 'product_id', indicator = True), I got a "ValueError". A search on Copilot indicated that the error message means that there’s a column name clash between the dataframes being merged. It mentioned that setting indicator=True tries to create a new column named _merge, but one of the dataframes already has a column with that same name exists in df_ords_prods_combined dataframe. The fix suggested was to give the indicator column a unique name. Therefore, I decided to drop it from the dataframe before operating the merging procedure.

In [21]:
# Drop unnecessary columns
df_ords_prods_combined = df_ords_prods_combined.drop(['_merge'], axis=1)

In [22]:
# Check the output
df_ords_prods_combined.head()

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0           0   2539329        1    prior             1          2   
1           0   2539329        1    prior             1          2   
2           0   2539329        1    prior             1          2   
3           0   2539329        1    prior             1          2   
4           0   2539329        1    prior             1          2   

   order_hour_of_day  days_since_prior_order  first_order  product_id  \
0                  8                     NaN         True         196   
1                  8                     NaN         True       14084   
2                  8                     NaN         True       12427   
3                  8                     NaN         True       26088   
4                  8                     NaN         True       26405   

   add_to_cart_order  reordered  
0                  1          0  
1                  2          0  
2                  3          0  
3                  4          0  
4                  5          0

In [23]:
# Check the output
df_ords_prods_combined.shape

(32434489, 12)

In [24]:
# Merge the updated dataframes
df_ords_prods_merge = df_prods.merge(df_ords_prods_combined, on = 'product_id', indicator = True)

In [24]:
# Check the output
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  first_order  add_to_cart_order  reordered _merge  
0                     3.0        False                  5          0   both  
1                    20.0        False                  1          1   both  
2                     6.0        False                 20          0   both  
3                     NaN         True                 10          0   both  
4                     9.0        False                 11          1   both

In [25]:
df_ords_prods_merge.shape

(32404859, 17)

### 4. Confirm the results of the merge using the merge flag

In [26]:
df_ords_prods_merge['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

#### The resulting dataframe (after the merge) has 32,404,859 rows, and each of those rows have information found in both input data sets, as we used an inner join for the purposes of this project.

### 5. Export the newly created dataframe as ords_prods_merge in a suitable format

In [27]:
# Export data to pkl
df_ords_prods_merge.to_pickle(os.path.join(path, 'Prepared Data', 'ords_prods_merge.pkl'))